event-handler/teleport_events_watcher_test.go

```diff
@@ -136,6 +138,142 @@ func TestNext(t *testing.T) {
+type errMockTeleportEventWatcher struct {
+	// events is the mock list of events
+	events []events.AuditEvent
+	t      *testing.T
+	called bool
+}
+
+// SearchEvents is mock SearchEvents method which returns events
+func (c *errMockTeleportEventWatcher) SearchEvents(ctx context.Context, fromUTC, toUTC time.Time, namespace string, eventTypes []string, limit int, order types.EventOrder, startKey string) ([]events.AuditEvent, string, error) {
+	e := c.events
+	c.events = make([]events.AuditEvent, 0) // nullify events
+	return e, "test", nil
+}
+
+// StreamSessionEvents returns session events stream
+func (c *errMockTeleportEventWatcher) StreamSessionEvents(ctx context.Context, sessionID string, startIndex int64) (chan events.AuditEvent, chan error) {
+	return nil, nil
+}
+
+// SearchEvents is mock SearchEvents method which returns events
+func (c *errMockTeleportEventWatcher) SearchUnstructuredEvents(ctx context.Context, fromUTC, toUTC time.Time, namespace string, eventTypes []string, limit int, order types.EventOrder, startKey string) ([]*auditlogpb.EventUnstructured, string, error) {
+	if !c.called {
+		e := c.events
+		c.events = make([]events.AuditEvent, 0) // nullify events
+
+		events := make([]*auditlogpb.EventUnstructured, len(e))
+		for i, event := range e {
+			events[i] = eventToJSON(c.t, event)
+		}
+		c.called = true
+		return events, "", nil
+	}
+
+	return nil, "", errors.New("error")
+}
+
+// StreamSessionEvents returns session events stream
+func (c *errMockTeleportEventWatcher) StreamUnstructuredSessionEvents(ctx context.Context, sessionID string, startIndex int64) (chan *auditlogpb.EventUnstructured, chan error) {
+	return nil, nil
+}
+
+// UsertLock is mock UpsertLock method
+func (c *errMockTeleportEventWatcher) UpsertLock(ctx context.Context, lock types.Lock) error {
+	return nil
+}
+
+func (c *errMockTeleportEventWatcher) Ping(ctx context.Context) (proto.PingResponse, error) {
+	return proto.PingResponse{
+		ServerVersion: Version,
+	}, nil
+}
+
+// Close is mock close method
+func (c *errMockTeleportEventWatcher) Close() error {
+	return nil
+}
+
+func newErrTeleportEventWatcher(t *testing.T, e []events.AuditEvent, exitOnLastEvent bool) *TeleportEventsWatcher {
+	teleportEventWatcher := &errMockTeleportEventWatcher{events: e, t: t}
+
+	client := &TeleportEventsWatcher{
+		client: teleportEventWatcher,
+		pos:    -1,
+		config: &StartCmdConfig{
+			IngestConfig: IngestConfig{
+				BatchSize:       5,
+				ExitOnLastEvent: exitOnLastEvent,
+			},
+		},
+	}
+
+	return client
+}
+
+func TestLastEvent(t *testing.T) {
+	t.Run("should not leave hanging go-routines", func(t *testing.T) {
+		const (
+			nIters = 5
+		)
+
+		e := []events.AuditEvent{
+			&events.UserCreate{
+				Metadata: events.Metadata{
+					ID: "1",
+				},
+			},
+		}
+
+		client := newErrTeleportEventWatcher(t, e, false)
+
+		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+		defer cancel()
+
+		t.Cleanup(cancel)
```

Comment:
You don't need to call cancel twice. Either one of these is fine, so pick one and get rid of the other.